lib/burnside/mixin.h

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_BURNSIDE_MIXIN_H_
+#define FORTRAN_BURNSIDE_MIXIN_H_
+
+// Mixin classes are "partial" classes (not used standalone) that can be used to
+// add a repetitive (ad hoc) interface (and implementation) to a class.  It's
+// better to think of these as "included in" a class, rather than as an
+// "inherited from" base class.
+
+#include "llvm/ADT/ilist.h"
+#include <optional>
+#include <tuple>
+#include <type_traits>
+#include <variant>
+
+namespace Fortran::burnside {
+
+// implementation of a (moveable) sum type (variant)
```

Comment:
This template class has only one use, which perhaps could be changed to `SumTypeCopyMixin`.